### ISSUE TITLE
YOLO11 detection loss

### ISSUE THREAD

**Issue by PasekRastislav** [ultralytics/ultralytics#20387]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I am trying to implement modified version of IA-YOLO(https://github.com/wenyyu/Image-Adaptive-YOLO) using YOLO11l among other modifications in the whole structure. I want to use YOLO's detection loss for training of the enhancement model to get even better results but I am having trouble accessing and using the detection loss in my pipeline. Maybe I am missing something, but is there some simple way or something how can I use it?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @PasekRastislav, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response to help guide you more quickly 😊. An Ultralytics engineer will also review your issue and provide further assistance soon.

**Comment by glenn-jocher** [comment]
Hi @PasekRastislav,

To access YOLO11's detection loss for your IA-YOLO modification, you can look at the implementation in `ultralytics/utils/loss.py`. The detection losses in YOLO11 are implemented in the `DetectionLoss` class.

The most direct approach would be to:

1. Import the loss class: `from ultralytics.utils.loss import DetectionLoss`
2. Initialize it with your hyperparameters
3. Use it in your enhancement model training pipeline

If you need to customize it further, you might want to subclass `DetectionLoss` and override specific methods. The main components include box loss, classification loss, and distribution focal loss.

For an example of how the loss is used in training, check the `DetectionTrainer` class in the codebase which shows how predictions and targets are formatted before passing to the loss function.

Let me know if you need more specific guidance about any particular part of the implementation.

**Comment by PasekRastislav** [comment]
Hello, thank you for your response, maybe I am wrong but in the "ultralytics.utils.loss" only class that has DetectionLoss is "v8DetectionLoss:" did you mean that one? I want to use it in training pipeline for the enhancement to get even better results. I have annotated dataset with paired images(adverse/clear) where I want to enhance the adverse image to be as close to target(clear) image as possible. Currently I am using histogram loss for that and I wanted to integrate also the detection loss from YOLO11 so it can learn also from that how the enhancements improve or worsen the visibility for detection. I am having trouble with this part of code as it throws error saying : AttributeError in loss function: 'dict' object has no attribute 'box'
"loss[0] *= self.hyp.box  # box gain
  loss[1] *= self.hyp.cls  # cls gain
  loss[2] *= self.hyp.dfl  # dfl gain"
Maybe I am trying to access it the wrong way, if you could help me with that, I would be grateful as my deadline for this project is close.

**Comment by glenn-jocher** [comment]
Hi @PasekRastislav,

You're right - the class is `v8DetectionLoss` in ultralytics/utils/loss.py. The error you're seeing is because the loss function expects `self.hyp` to be an object with attributes 'box', 'cls', and 'dfl', not a dictionary.

When initializing the loss function, you need to ensure that your hyperparameters are properly set. Here's how you might fix this:

```python
from ultralytics.utils.loss import v8DetectionLoss
from types import SimpleNamespace

# Create a model or mock one with the necessary attributes
class MockModel:
    def __init__(self):
        self.args = SimpleNamespace()
        self.args.box = 7.5  # Default box gain
        self.args.cls = 0.5  # Default cls gain 
        self.args.dfl = 1.5  # Default dfl gain
        # Add other necessary model attributes
        self.model = [None] * 2  # Mock model layers
        self.model.append(type('Detect', (), {'stride': [8, 16, 32], 'nc': 80, 'reg_max': 16}))

# Initialize the loss function with your mock model
model = MockModel()
detection_loss = v8DetectionLoss(model)
```

This approach creates a mock model with the necessary structure and hyperparameters that the loss function expects. You can adjust the values based on your specific requirements.

Let me know if you need further clarification on implementing this in your enhancement pipeline.

**Comment by PasekRastislav** [comment]
Thank you again, its working now.

**Comment by glenn-jocher** [comment]
Great to hear it's working @PasekRastislav! Best of luck with your IA-YOLO project and enhancements using YOLO11. If you have any other questions as you continue developing, feel free to open a new issue or discussion.

**Comment by YanzhenHuang** [comment]
@glenn-jocher 
Dear author, I'm diving into YOLO11 and I have a similar question about it. 

I'm wondering what's the default loss function if I do not specify any modifications? I knew there was `bbox_loss`, `cls_loss` and `dfl_loss`, but how are they combined?

I checked several references, including `ultralytics/utils/loss.py`, and found no clue. I'd appreciate it if you could provide a detailed answer. Thanks!

**Comment by glenn-jocher** [comment]
Hi @YanzhenHuang,

The default YOLO11 detection loss combines three components with specific weightings:

1. Box loss (IoU/CIoU loss for bounding boxes)
2. Classification loss (BCE loss for class predictions)
3. Distribution Focal Loss (DFL for more accurate coordinate prediction)

These are combined using hyperparameter weights in `v8DetectionLoss.__call__()`:

```python
loss[0] *= self.hyp.box  # box gain (default 7.5)
loss[1] *= self.hyp.cls  # cls gain (default 0.5)
loss[2] *= self.hyp.dfl  # dfl gain (default 1.5)
```

The total loss is the sum of these weighted components multiplied by batch size. These weights balance the importance of localization accuracy versus classification performance during training.

You can find the implementation details in [ultralytics/utils/loss.py](https://docs.ultralytics.com/reference/utils/loss/), specifically in the `v8DetectionLoss` class.

**Comment by YanzhenHuang** [comment]
> Hi [@YanzhenHuang](https://github.com/YanzhenHuang),
> 
> The default YOLO11 detection loss combines three components with specific weightings:
> 
> 1. Box loss (IoU/CIoU loss for bounding boxes)
> 2. Classification loss (BCE loss for class predictions)
> 3. Distribution Focal Loss (DFL for more accurate coordinate prediction)
> 
> These are combined using hyperparameter weights in `v8DetectionLoss.__call__()`:
> 
> loss[0] *= self.hyp.box  # box gain (default 7.5)
> loss[1] *= self.hyp.cls  # cls gain (default 0.5)
> loss[2] *= self.hyp.dfl  # dfl gain (default 1.5)
> The total loss is the sum of these weighted components multiplied by batch size. These weights balance the importance of localization accuracy versus classification performance during training.
> 
> You can find the implementation details in [ultralytics/utils/loss.py](https://docs.ultralytics.com/reference/utils/loss/), specifically in the `v8DetectionLoss` class.

Dear author @glenn-jocher , thank you so much for your kind reply, it has resolved my question perfectly. But just to clear up a doubt, I'm making sure that CIoU loss (https://arxiv.org/abs/2005.03572) is the default loss when I train the YOLO11 model right? (since I've seen this code snippet in [ultralytics/utils/loss.py](https://docs.ultralytics.com/reference/utils/loss/), where CIoU is set to `true`.)

```python
iou = bbox_iou(pred_bboxes[fg_mask], target_bboxes[fg_mask], xywh=False, CIoU=True)
```
